File: CharGen/CharGen/STOGen.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CharGen
{
    class STOGen
    {
        // Declare the arrays
        private String[] factions;
        private String[] races;
        private String[] charClasses;
        private String[] shipClasses;

        // And the randomizer
        private Random rand;
        private int randomValue;

        // And finally a space to hold the character
        public STOChar character;
        // Constructor to populate the faction and character class arrays
        public STOGen()
        {
            // Initialize the randomizer
            rand = new Random();

            // Declare the arrays
            factions = new String[5];
            races = new String[13];
            charClasses = new String[3];
            shipClasses = new String[3];

            // Populate the arrays
            factions[0] = "Federation";
            factions[1] = "Klingon";
            factions[2] = "Romulan (Federation aligned)";
            factions[3] = "Romulan (Klingon aligned)";
            factions[4] = "TOS Federation";
            charClasses[0] = "Tactical";
            charClasses[1] = "Science";
            charClasses[2] = "Engineering";

            // As ship classes are different by faction, ship class array will be populated once faction is randomly determined.

            // And initialize the holding variable
            character = new STOChar();
        }

        public int populateShipClasses(char faction) // F for Fed, K for KDF, R for Romulan.  Returns the number of possible classes that have ships from T2 and up.
        {
            try
            {

                switch (faction)
                {
                    case 'F':
                        shipClasses[0] = "Cruiser";
                        shipClasses[1] = "Escort";
                        shipClasses[2] = "Science Vessel";
                        return 3;
                    case 'K':
                        shipClasses[0] = "Raider";
                        shipClasses[1] = "Raptor";
                        shipClasses[2] = "Battle Cruiser";
                        return 3;
                    case 'R':
                        shipClasses[0] = "Warbird";
                        // There's only one ship class that meets the specifications.
                        shipClasses[1] = null;
                        shipClasses[2] = null;
                        return 1;
                    default:
                        throw new Exception();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Cannot populate ship classes: Not a valid faction.");
                return -1;
            }
        }

        public int populateRaces(int faction) // Int value corresponds to the factions array.  Returns the number of possible races, NOT counting C-Store or Lifetime races, or the Alien race.
        {
            try
            {
                switch(faction)
                {
                    case 0: // Federation
                        races[0] = "Andorian";
                        races[1] = "Bajoran";
                        races[2] = "Benzite";
                        races[3] = "Betazoid";
                        races[4] = "Bolian";
                        races[5] = "Ferengi";
                        races[6] = "Human";
                        races[7] = "Pakled";
                        races[8] = "Rigelian";
                        races[9] = "Saurian";
                        races[10] = "Tellarite";
                        races[11] = "Trill";
                        races[12] = "Vulcan";
                        return 13;
                    case 1: // KDF
                        races[0] = "Gorn";
                        races[1] = "Klingon";
                        races[2] = "Lethean";
                        races[3] = "Nausicaan";
                        races[4] = "Orion";
                        // Only five races available to KDF
                        races[5] = null;
                        races[6] = null;
                        races[7] = null;
                        races[8] = null;
                        races[9] = null;
                        races[10] = null;
                        races[11] = null;
                        races[12] = null;
                        return 5;
                    case 2:
                    case 3: // Romulans
                        races[0] = "Romulan";
                        // Only Romulans available
                        races[1] = null;
                        races[2] = null;
                        races[3] = null;
                        races[4] = null;
                        races[5] = null;
                        races[6] = null;
                        races[7] = null;
                        races[8] = null;
                        races[9] = null;
                        races[10] = null;
                        races[11] = null;
                        races[12] = null;
                        return 1;
                    case 4: // TOS Federation
                        races[0] = "23c Human";
                        races[1] = "23c Andorian";
                        races[2] = "23c Tellarite";
                        races[3] = "23c Vulcan";
                        // Only four races available to TOS Federation
                        races[4] = null;
                        races[5] = null;
                        races[6] = null;
                        races[7] = null;
                        races[8] = null;
                        races[9] = null;
                        races[10] = null;
                        races[11] = null;
                        races[12] = null;
                        return 4;
                    default:
                        throw new Exception();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Cannot populate ship classes: Not a valid faction.");
                return -1;
            }

        }

        public void randomCharacter() // Random faction
        {
            // Randomize faction
            randomValue = rand.Next(5);
            int faction = randomValue; // Will be used for ship selection
            character.faction = factions[randomValue];

            // Randomize character class
            randomValue = rand.Next(3);
            character.charClass = charClasses[randomValue];

            // Determine which faction's races
            int numRaces;
            numRaces = populateRaces(faction);

            // Randomize race
            randomValue = rand.Next(numRaces);
            character.race = races[randomValue];

            // Determine which faction's ships
            int numShips;
            try
            {
                switch (faction)
                {
                    // Federation ships
                    case 0:
                    case 4:
                        numShips = populateShipClasses('F');
                        break;
                    // Klingon ships
                    case 1:
                        numShips = populateShipClasses('K');
                        break;
                    // Romulan ships
                    case 2:
                    case 3:
                        numShips = populateShipClasses('R');
                        break;
                    default:
                        throw new Exception();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Cannot select ship faction: Faction number outside range.");
                return;
            }

            // Randomize ship class
            randomValue = rand.Next(numShips);
            character.shipClass = shipClasses[randomValue];
        }

        public void randomCharacter(char alignment) // Specific alignment.  F for Fed, K for KDF
        {
            // Randomize faction within alignment
            if (alignment == 'F') // Fed alignment
            {
                do
                {
                    randomValue = rand.Next(5);
                } while (randomValue != 0 && randomValue != 2 && randomValue != 4); // Throw away values that lead to KDF alignment
            }
            else if(alignment == 'K') // KDF alignment
            {
                do
                {
                    randomValue = rand.Next(5);
                } while (randomValue != 1 && randomValue != 3); // Throw away values that lead to Fed alignment
            }
            else
            {
                Console.WriteLine("Invalid alignment.");
                return;
            }

            int faction = randomValue; // Will be used for ship selection
            character.faction = factions[randomValue];

            // Determine which faction's races
            int numRaces;
            numRaces = populateRaces(faction);

            // Randomize race
            randomValue = rand.Next(numRaces);
            character.race = races[randomValue];
            
            // Randomize character class
            randomValue = rand.Next(3);
            character.charClass = charClasses[randomValue];

            // Determine which faction's ships
            int numShips;
            try
            {
                switch (faction)
                {
                    // Federation ships
                    case 0:
                    case 4:
                        numShips = populateShipClasses('F');
                        break;
                    // Klingon ships
                    case 1:
                        numShips = populateShipClasses('K');
                        break;
                    // Romulan ships
                    case 2:
                    case 3:
                        numShips = populateShipClasses('R');
                        break;
                    default:
                        throw new Exception();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Cannot select ship faction: Faction number outside range.");
                return;
            }

            // Randomize ship class
            randomValue = rand.Next(numShips);
            character.shipClass = shipClasses[randomValue];
        }

        public void randomCharacter(int factionPicked) // Specific faction.
        {
            int faction = factionPicked; // Will be used for ship selection
            character.faction = factions[factionPicked];

            // Determine which faction's races
            int numRaces;
            numRaces = populateRaces(faction);

            // Randomize race
            randomValue = rand.Next(numRaces);
            character.race = races[randomValue];
            
            // Randomize character class
            randomValue = rand.Next(3);
            character.charClass = charClasses[randomValue];

            // Determine which faction's ships
            int numShips;
            try
            {
                switch (faction)
                {
                    // Federation ships
                    case 0:
                    case 4:
                        numShips = populateShipClasses('F');
                        break;
                    // Klingon ships
                    case 1:
                        numShips = populateShipClasses('K');
                        break;
                    // Romulan ships
                    case 2:
                    case 3:
                        numShips = populateShipClasses('R');
                        break;
                    default:
                        throw new Exception();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Cannot select ship faction: Faction number outside range.");
                return;
            }

            // Randomize ship class
            randomValue = rand.Next(numShips);
            character.shipClass = shipClasses[randomValue];
        }
    }
}

File: CharGen2/src/charGen2/STOGen.java
package charGen2;

import java.util.Random;

public class STOGen {
	// Declare the arrays
	private String[] factions;
	private String[] races;
	private String[] charClasses;
	private String[] shipClasses;
	
	// And the randomizer
	private Random rand;
	private int randomValue;
	
	// Flags
	private boolean dominion;
	private int zen;
	private boolean lifetime;
	private boolean reman;
	private String side;
	
	public boolean isDominion() {
		return dominion;
	}

	public void setDominion(boolean dominion) {
		this.dominion = dominion;
	}

	public int getZen() {
		return zen;
	}

	public void setZen(int zen) {
		this.zen = zen;
	}

	public boolean isLifetime() {
		return lifetime;
	}

	public void setLifetime(boolean lifetime) {
		this.lifetime = lifetime;
	}

	// And finally a space to hold the character
	private STOChar character;
	
	public STOChar getCharacter() {
		return character;
	}

	public void setCharacter(STOChar character) {
		this.character = character;
	}

	// Constructor to populate the faction and character class arrays
	public STOGen()
	{
		// Initialize the randomizer
		rand = new Random();
		
		// Declare the arrays
		factions = new String[8];
		races = new String[20];
		charClasses = new String[3];
		shipClasses = new String[3];
		
		// Populate the static array
		charClasses[0] = "Tactical";
		charClasses[1] = "Science";
		charClasses[2] = "Engineering";
		
		// As ship classes are different by faction, ship class array will be populated once faction is randomly determined.
		
		// And initialize the holding variable
		character = new STOChar();
	}
	
	public int populateFactions(String side)
	{
		if(side == "Fed") {
			factions[0] = "Federation";
			factions[1] = "Romulan (Federation aligned)";
			factions[2] = "TOS Federation";
			factions[3] = "DSC Federation";
		
			if(dominion) {
				factions[4] = "Dominion (Federation aligned)";
				return 5;
			}
			else {
				return 4;
			}
		}
		else if(side == "KDF") {
			factions[0] = "Klingon";
			factions[1] = "Romulan (Klingon aligned)";
		
			if(dominion) {
				factions[2] = "Dominion (Klingon aligned)";
				return 3;
			}
			else {
				return 2;
			}
		}
		else {
			factions[0] = "Federation";
			factions[1] = "Klingon";
			factions[2] = "Romulan (Federation aligned)";
			factions[3] = "Romulan (Klingon aligned)";
			factions[4] = "TOS Federation";
			factions[5] = "DSC Federation";
		
			if(dominion) {
				factions[6] = "Dominion (Federation aligned)";
				factions[7] = "Dominion (Klingon aligned)";
				return 8;
			}
			else {
				return 6;
			}
		}

	}
	
	public int populateShipClasses(String faction) // F for Fed, K for KDF, R for Romulan, D for Dominion.  Returns the number of possible classes that have ships from T2 and up.
	{
		try{
			
			switch(faction)
			{
			case "Fed":
				shipClasses[0] = "Cruiser";
				shipClasses[1] = "Escort";
				shipClasses[2] = "Science Vessel";
				return 3;
			case "KDF":
				shipClasses[0] = "Raider";
				shipClasses[1] = "Raptor";	
				shipClasses[2] = "Battle Cruiser";
				return 3;
			case "Rom":
				shipClasses[0] = "Warbird";
				// There's only one ship class that meets the specifications.
				return 1;
			case "Dom":
				shipClasses[0] = "Escort";
				// There's only one ship class that meets the specifications.
				return 1;
			default:
				throw new IllegalArgumentException();
			}
		}
		catch(IllegalArgumentException e)
		{
			System.out.println("Cannot populate ship classes: Not a valid faction.");
			return -1;
		}
	}
	
	public int populateRaces(String faction) //Returns the number of possible races, given the max amount of ZEN
	{
		try{
			int currentIndex = 0;
			switch(faction)
			{
			case "Fed":
				races[currentIndex++] = "Andorian";
				races[currentIndex++] = "Bajoran";
				races[currentIndex++] = "Benzite";
				races[currentIndex++] = "Betazoid";
				races[currentIndex++] = "Bolian";
				races[currentIndex++] = "Ferengi";
				races[currentIndex++] = "Human";
				races[currentIndex++] = "Pakled";
				races[currentIndex++] = "Rigelian";
				races[currentIndex++] = "Saurian";
				races[currentIndex++] = "Tellarite";
				races[currentIndex++] = "Trill";
				races[currentIndex++] = "Vulcan";
				races[currentIndex++] = "Alien";
				if(zen >= 600) {
					races[currentIndex++] = "Caitian";
					races[currentIndex++] = "Klingon";
					races[currentIndex++] = "Trill (joined)";

				}
				if(zen >= 1000) {
					races[currentIndex++] = "Cardassian";
				}
				if(lifetime) {
					races[currentIndex++] = "Liberated Borg";
					races[currentIndex++] = "Talaxian";
				}
				return currentIndex;
			case "TOS":
				races[currentIndex++] = "Andorian";
				races[currentIndex++] = "Human";
				races[currentIndex++] = "Tellarite";
				races[currentIndex++] = "Vulcan";
				return currentIndex;
			case "DSC":
				races[currentIndex++] = "Human";
				races[currentIndex++] = "Vulcan";
				races[currentIndex++] = "Alien";
				return currentIndex;
			case "KDF":
				races[currentIndex++] = "Gorn";
				races[currentIndex++] = "Klingon";
				races[currentIndex++] = "Lethean";
				races[currentIndex++] = "Nausicaan";
				races[currentIndex++] = "Orion";
				races[currentIndex++] = "Alien";
				if(zen >= 600) {
					races[currentIndex++] = "Ferasan";
					races[currentIndex++] = "Trill (joined)";
				}
				if(zen >= 1000) {
					races[currentIndex++] = "Cardassian";
				}
				if(lifetime) {
					races[currentIndex++] = "Liberated Borg";
					races[currentIndex++] = "Talaxian";
				}
				return currentIndex;
			case "Rom":
				races[currentIndex++] = "Romulan";
				races[currentIndex++] = "Alien";
				if(lifetime) {
					races[currentIndex++] = "Liberated Borg";
				}
				if(reman) {
					races[currentIndex++] = "Reman";
				}
				return currentIndex;
			case "Dom":
				races[currentIndex++] = "Jem'Hadar";
				if(zen >= 600) {
					races[currentIndex++] = "Jem'Hadar Vanguard";
				}
				return currentIndex;
			default:
				throw new IllegalArgumentException();
			}
		}
		catch(IllegalArgumentException e)
		{
			System.out.println("Cannot populate races: Not a valid faction.");
			return -1;
		}
	}
	
	public void randomCharacter()
	{
		// Randomize faction
		int numFactions = populateFactions(side);
		randomValue = rand.nextInt(numFactions);
		character.setFaction(factions[randomValue]);
		
		// Randomize gender
		if(character.getFaction().startsWith("Dominion")) {
			character.setGender("Male"); // Dominion characters are always male
		}
		else {
			randomValue = rand.nextInt(2);
			if(randomValue == 0) {
				character.setGender("Male");
			}
			else {
				character.setGender("Female");
			}
		}
		
		// Randomize character class
		randomValue = rand.nextInt(3);
		character.setCharClass(charClasses[randomValue]);
		
		// Determine each faction's races
		int numRaces;
		try {
			switch(character.getFaction())
			{
			// Fed
			case "Federation":
				numRaces = populateRaces("Fed");
				break;
			// KDF
			case "Klingon":
				numRaces = populateRaces("KDF");
				break;
			// Romulan
			case "Romulan (Federation aligned)":
			case "Romulan (Klingon aligned)":
				numRaces = populateRaces("Rom");
				break;
			// TOS
			case "TOS Federation":
				numRaces = populateRaces("TOS");
				break;
			// DSC
			case "DSC Federation":
				numRaces = populateRaces("DSC");
				break;
			// Dominion
			case "Dominion (Federation aligned)":
			case "Dominion (Klingon aligned)":
				numRaces = populateRaces("Dom");
				break;
			default:
				throw new IllegalArgumentException();
			}
		}
		catch(IllegalArgumentException e)
		{
			System.out.println("Cannot select ship faction: Faction number outside range.");
			return;
		}
		
		// Randomize race
		randomValue = rand.nextInt(numRaces);
		character.setRace(races[randomValue]);
		
		
		// Determine which faction's ships
		int numShips;
		try{
			switch(character.getFaction())
			{
			// Federation ships
			case "Federation":
			case "TOS Federation":
			case "DSC Federation":
				numShips = populateShipClasses("Fed");
				break;
			// Klingon ships
			case "Klingon":
				numShips = populateShipClasses("KDF");
				break;
			// Romulan ships
			case "Romulan (Federation aligned)":
			case "Romulan (Klingon aligned)":
				numShips = populateShipClasses("Rom");
				break;
			// Dominion ships
			case "Dominion (Federation aligned)":
			case "Dominion (Klingon aligned)":
				numShips = populateShipClasses("Dom");
				break;
			default:
				throw new IllegalArgumentException();
			}
		}
		catch(IllegalArgumentException e)
		{
			System.out.println("Cannot select ship faction: Faction number outside range.");
			return;
		}
		
		// Randomize ship class
		randomValue = rand.nextInt(numShips);
		character.setShipClass(shipClasses[randomValue]);
	}
	
	public void printCharacter()
	{
		System.out.println("STO character generated.");
		System.out.println("Faction: " + character.getFaction());
		System.out.println("Race: " + character.getRace());
		System.out.println("Gender: " + character.getGender());
		System.out.println("Character class: " + character.getCharClass());
		System.out.println("Ship class: " + character.getShipClass());
	}
	
	public void generate()
	{
		char input;
		boolean properInput;
		try {
			do {
				System.out.println();
				System.out.print("Do you have a Lifetime subscription?  (Y/N) ");
				input = (char) System.in.read();
				switch(input)
				{
				case 'Y':
				case 'y':
					lifetime = true;
					properInput = true;
					break;
				case 'N':
				case 'n':
					lifetime = false;
					properInput = true;
					break;
				default:
					System.out.println();
					System.out.println("Invalid character.");
					properInput = false;
					break;
				}
			} while(properInput == false);
		}
		catch(Exception e)
		{
			System.out.println("Error reading keypress.");
		}

		try {
			do {
				System.out.println();
				System.out.print("Have you unlocked the Reman race?  (Y/N) ");
				input = (char) System.in.read();
				switch(input)
				{
				case 'Y':
				case 'y':
					reman = true;
					properInput = true;
					break;
				case 'N':
				case 'n':
					reman = false;
					properInput = true;
					break;
				default:
					System.out.println();
					System.out.println("Invalid character.");
					properInput = false;
					break;
				}
			} while(properInput == false);
		}
		catch(Exception e)
		{
			System.out.println("Error reading keypress.");
		}
		
		try {
			do {
				System.out.println();
				System.out.print("Do you want to include Dominion characters in the generator table?  (Y/N) ");
				input = (char) System.in.read();
				switch(input)
				{
				case 'Y':
				case 'y':
					dominion = true;
					properInput = true;
					break;
				case 'N':
				case 'n':
					dominion = false;
					properInput = true;
					break;
				default:
					System.out.println();
					System.out.println("Invalid character.");
					properInput = false;
					break;
				}
			} while(properInput == false);
		}
		catch(Exception e)
		{
			System.out.println("Error reading keypress.");
		}
		
		try {
			do {
				System.out.println();
				System.out.print("Do you want to include (F)ederation-side characters, (K)lingon-side characters, or (B)oth? ");
				input = (char) System.in.read();
				switch(input)
				{
				case 'F':
				case 'f':
					side = "Fed";
					properInput = true;
					break;
				case 'K':
				case 'k':
					side = "KDF";
					properInput = true;
					break;
				case 'B':
				case 'b':
					side = "All";
					properInput = true;
					break;
				default:
					System.out.println();
					System.out.println("Invalid character.");
					properInput = false;
					break;
				}
			} while(properInput == false);
		}
		catch(Exception e)
		{
			System.out.println("Error reading keypress.");
		}
		
		
		try {
			System.out.println();
			System.out.print("What is the maximum number of ZEN you would be willing to spend to unlock a character race? ");
			zen = (int)System.in.read();
		}
		catch(Exception e)
		{
			System.out.println("Error reading amount.");
		}


		randomCharacter();
		printCharacter();
	}
}
